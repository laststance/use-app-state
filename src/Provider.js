// @flow // $FlowIssue flow doesn't know hook
import React, { useState, Children } from 'react'
import isPlainObject from './util/isPlainObject'
import Context from './Context'

type Props = {
  appState: Object,
  children: React$Node
}

export default function Provider(props: Props) {
  const { appState } = props

  if (!isPlainObject(appState)) {
    throw new Error('react-appstate: Provider Expected the appState to be a PlainObject')
  }

  // "tmpAppState" role is avoid name collision "appState"
  const [tmpAppState, setState] = useState(appState)

  const setAppState: Function = (state: Object) => {
    if (!isPlainObject(state)) {
      throw new Error(
        'react-appstate: Expected the setAppState() argument to be a PlainObject'
      )
    }

    setState(tmpAppState => {
      return { ...tmpAppState, ...state }
    })
  }

  return (
    <Context.Provider
      value={{
        appState: tmpAppState,
        setAppState: setAppState
      }}
    >
      {Children.only(props.children)}
    </Context.Provider>
  )
}
