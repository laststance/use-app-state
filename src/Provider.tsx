import React, { FC, useState, Children, ReactNode } from 'react'
import isPlainObject from './util/isPlainObject'
import Context from './Context'

// Originary type AppState = Record<string, unknown> but TypeScript doing incorrect compatibility evaluation.
// follow the link to detail https://github.com/microsoft/TypeScript/issues/15300
// eslint-disable-next-line @typescript-eslint/ban-types
export type AppState = Object
export type SetAppState = (state: AppState) => void

interface Props {
  initialState: AppState
  children: ReactNode
}

const Provider: FC<Props> = (props) => {
  const { initialState } = props

  if (!isPlainObject(initialState)) {
    throw new Error(
      'use-app-state: Provider Expected the appState to be a PlainObject'
    )
  }

  const [appState, setState] = useState(initialState)

  const setAppState: SetAppState = (newState: AppState) => {
    if (!isPlainObject(newState)) {
      throw new Error(
        'use-app-sate: Expected the setAppState() argument to be a PlainObject'
      )
    }

    setState(() => {
      return { ...appState, ...newState }
    })
  }

  return (
    <Context.Provider
      value={{
        appState: appState,
        setAppState: setAppState,
      }}
    >
      {Children.only(props.children)}
    </Context.Provider>
  )
}

export default Provider
