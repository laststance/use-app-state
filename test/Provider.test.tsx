import React from 'react'
import Provider from '../src/index'
import TestRenderer from 'react-test-renderer'

test(`doesn't work without children`, () => {
  expect(() =>
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore Because expectend used to JavaScirpt
    TestRenderer.create(<Provider initialState={{}} />)
  ).toThrowErrorMatchingSnapshot()
})

test(`doesn't work without giving appState`, () => {
  expect(() =>
    TestRenderer.create(
      // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
      // @ts-ignore Because expectend used to JavaScirpt
      <Provider>
        <div>elm</div>
      </Provider>
    )
  ).toThrowErrorMatchingSnapshot()
})

const renderElm = appState =>
  TestRenderer.create(
    <Provider initialState={appState}>
      <div>elm</div>
    </Provider>
  ).toJSON()

test(`work fine with PlainObject appState`, () => {
  const tree01 = renderElm({})
  expect(tree01).toMatchSnapshot()

  const tree02 = renderElm(new Object())
  expect(tree02).toMatchSnapshot()

  const tree03 = renderElm({ name: 'mark' })
  expect(tree03).toMatchSnapshot()

  const tree04 = renderElm({
    chunk: { home: { oneFloor: 'spacious', twoFloor: 'narrow' } },
    array: [1, 2, 3, 4]
  })
  expect(tree04).toMatchSnapshot()
})

test(`doesn't work without PlainObject`, () => {
  expect(() => renderElm(1)).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm([])).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm(new Map())).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm(new Date())).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm('string')).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm(true)).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm(undefined)).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm(null)).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm(NaN)).toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
  expect(() => renderElm({ count: 0 })).not.toThrowError(
    'use-app-state: Provider Expected the appState to be a PlainObject'
  )
})
