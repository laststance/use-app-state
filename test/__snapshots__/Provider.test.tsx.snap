// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't work without children 1`] = `"React.Children.only expected to receive a single React element child."`;

exports[`doesn't work without giving appState 1`] = `"react-appstate: Provider Expected the appState to be a PlainObject"`;

exports[`work fine with PlainObject appState 1`] = `
<div>
  elm
</div>
`;

exports[`work fine with PlainObject appState 2`] = `
<div>
  elm
</div>
`;

exports[`work fine with PlainObject appState 3`] = `
<div>
  elm
</div>
`;

exports[`work fine with PlainObject appState 4`] = `
<div>
  elm
</div>
`;
